ixdiagnose/plugins/zfs.py

```diff
@@ -0,0 +1,172 @@
+                output += zfs_getacl(resource_context, prop_dict)
+
+            prop_dict = {}
+            header_str = f'{resource_type} get all {resource_name}'
+            next_line = '\n\n' if index != 0 else ''
+            output += f'{next_line}{"=" * (len(header_str) + 5)}\n  {header_str}  \n{"=" * (len(header_str) + 5)}\n\n'
+            output += f'{props_header}\n'
+            resource_context = resource_name
+
+        if resource_type == 'zfs':
+            prop = resource_line.split()[1]
+            if prop in prop_list:
+                prop_dict[prop] = resource_line.split()[2]
+
+        output += f'{resource_line}\n'
+
+    if resource_name is not None and resource_type == 'zfs':
+        output += zfs_getacl(resource_name, prop_dict)
+
+    return output
+
+
+def encryption_summary(client: MiddlewareClient, context: Any) -> str:
+    summary = {}
+    for pool in client.call('pool.query'):
+        summary[pool['name']] = client.call('pool.dataset.encryption_summary', pool['name'], job=True)
+
+    return dumps(summary)
+
+
+def kstat_output(client: MiddlewareClient, context: Any) -> str:
+    kstats = ['fletcher_4_bench', 'vdev_raidz_bench', 'dbgmsg']
+    for pool in map(lambda p: p['name'], client.call('zfs.pool.query')):
+        kstats.extend([
+            f'{pool}/multihost',
+            f'{pool}/state',
+            f'{pool}/txgs',
+        ])
+
+    output = ''
+    for index, kstat in enumerate(kstats):
+        kstat_header = f'kstat {kstat!r}'
+        newline = '\n\n' if index != 0 else ''
+        header = f'{newline}{"=" * (len(kstat_header) + 5)}\n  {kstat_header}  \n{"=" * (len(kstat_header) + 5)}\n'
+        output += header
+        kstat_path = os.path.join('/proc/spl/kstat/zfs', kstat)
+        if os.path.exists(kstat_path):
+            with open(kstat_path, 'r') as f:
+                data = f.read()
+            output += f'\n{data}\n'
+        else:
+            output += f'\n{kstat_path!r} does not exist\n'
+
+    return output
+
+
+class ZFS(Plugin):
+    name = 'zfs'
+    metrics = [
+        CommandMetric(
+            'dataset_list', [
+                Command(['zfs', 'list', '-ro', 'space,refer,mountpoint'], 'ZFS Pool(s)', serializeable=False),
+            ]
+        ),
+        CommandMetric(
+            'pool_list', [Command(['zpool', 'list', '-v'], 'ZFS Pool(s)', serializeable=False)]
+        ),
+        CommandMetric(
+            'pool_status', [Command(['zpool', 'status', '-v'], 'ZFS Pool(s) Status', serializeable=False)]
+        ),
+        CommandMetric('pool_history', [Command(['zpool', 'history'], 'ZFS Pool(s) History', serializeable=False)]),
+        MiddlewareClientMetric(
+            'pool_scrub_tasks', [
+                MiddlewareCommand('pool.scrub.query', result_key='scrub_tasks'),
+            ]
+        ),
+        MiddlewareClientMetric(
+            'replication', [
+                MiddlewareCommand('replication.query'),
```

Comment:
Non-LOCAL replication query joins SSH keys which will get exposed here.

Reply:
Let's move this and snapshot tasks to a separate module, ZFS report is long enough